src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -110,31 +110,46 @@ public static string GetAsciiOrUTF8StringNonNullCharacters(this ReadOnlySpan<byt
 
         public static string GetRequestHeaderString(this ReadOnlySpan<byte> span, string name, Func<string, Encoding?> encodingSelector)
         {
+            string result;
             if (ReferenceEquals(KestrelServerOptions.DefaultHeaderEncodingSelector, encodingSelector))
             {
-                return span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
+                result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
             }
-
-            var encoding = encodingSelector(name);
-
-            if (encoding is null)
+            else
```

Comment:
I'd split the `else`-part into a separate method, so that the path with the default encoding selector may be inlined.

Reply:
I think I'm missing something here, can you show me what you mean? It's not obvious to me how that's functionally different

Reply:
> how that's functionally different

It's just about performance for the default code-path, which is then shorter and may get inlined by the JIT.

Basically
```c#
            string result;
            if (ReferenceEquals(KestrelServerOptions.DefaultHeaderEncodingSelector, encodingSelector))
            {
                result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
            }
            else
            {
                result = GetRequestHeaderStringWithoutDefaultEncodingCore(...);  // or any better name for the method
            }

            // Non-tab control characters are considered invalid    
            ...
            // rest of code as is
```
So the idea is to streamline the code-path for the default-encoding, and move the code from that `else`-branch to a separate method.
Together with JIT-improvements like PGO, etc.* I hope this will give a littel boost to perf for the default encoding.

\* TBH right now I need to build up a feeling for these improvements, so it's not so easy to be quite sure what change in code has which effect in generated machine-code.